cpp/ql/src/experimental/Security/CWE/CWE-552/DangerousUseRename.ql

```diff
@@ -0,0 +1,58 @@
+/**
+ * @name Dangerous use of rename
+ * @description If the rename function did not work correctly, attempting to copy the contents of a source file by opening the target file without assessing permissions creates a mechanism for deleting an arbitrary system file.
+ * @kind problem
+ * @id cpp/dangerous-manual-copy-after-rename
+ * @problem.severity warning
+ * @precision medium
+ * @tags correctness
+ *       security
+ *       external/cwe/cwe-552
+ */
+
+import cpp
+import semmle.code.cpp.valuenumbering.GlobalValueNumbering
+
+/** Holds if the function argument is used in opening the file for reading or writing. */
+predicate findFileForReadOrWrite(FunctionCall fc, FunctionCall fc1, string mode, int na) {
+  fc1.getTarget().hasName(["open", "fopen"]) and
+  globalValueNumber(fc1.getArgument(0)) = globalValueNumber(fc.getArgument(na)) and
+  fc1.getArgument(1).getValue() = mode and
+  fc.getASuccessor+() = fc1
+}
+
+from FunctionCall renameCall
+where
+  renameCall.getTarget().hasName("rename") and
+  exists(IfStmt ifst, Expr ec, Expr ecd, FunctionCall readCall, FunctionCall writeCall |
+    findFileForReadOrWrite(renameCall, readCall, "r", 0) and
+    findFileForReadOrWrite(renameCall, writeCall, "w", 1) and
+    ecd = ifst.getCondition().getAChild*() and
+    (
+      globalValueNumber(ecd) = globalValueNumber(renameCall) and
+      not ecd.getParent() instanceof ComparisonOperation and
+      not ecd.getParent() instanceof NotExpr
+      or
+      exists(Expr evp |
+        globalValueNumber(evp) = globalValueNumber(renameCall) and
+        ecd.(ComparisonOperation).hasOperands(evp, _)
+      )
+    ) and
+    (
+      ec.getValue() = "0" and
+      ecd.(EQExpr).hasOperands(_, ec) and
+      forall(Expr st | st = ifst.getThen().getASuccessor*() | st != readCall) and
+      forall(Expr st | st = ifst.getThen().getASuccessor*() | st != writeCall)
```

Comment:
Which of your test cases covers this?

Reply:
`test5` `test6` `test7` `test8`

Reply:
I don't see this. If we look at `test5` for example:
```
  if (rename(from,to)==0) // BAD
    return;
  f1 = fopen(from, "r");
  count = fread(data, 1, 1000, f1);
  fclose(f1);
  remove(to);
  f2 = fopen(to, "w");
  fwrite(data, count, 1, f2);
  fclose(f2);
```
Then, `ifst.getThen().getASuccessor*()` gives you all the statements and expressions textually lower than `if (rename(from,to)==0)`, including `fopen(from, "r")` and `fopen(to, "w")`, so the `forall` cannot hold.